satellite/audit/service2.go

```diff
@@ -0,0 +1,102 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package audit
+
+import (
+	"context"
+	"time"
+
+	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
+)
+
+// Service2 contains information for populating audit queue and processing audits.
+type Service2 struct {
+	log    *zap.Logger
+	closed chan struct{}
+
+	workers []*worker
+	queue   *queue
+}
+
+// NewService2 instantiates Service2 and workers.
+func NewService2(log *zap.Logger, config Config) (*Service2, error) {
+	queue := &queue{}
+	var workers []*worker
+	for i := 0; i < config.WorkerCount; i++ {
+		workers = append(workers, newWorker(queue, config.QueueInterval))
+	}
+	return &Service2{
+		log:    log,
+		closed: make(chan struct{}),
+
+		workers: workers,
+		queue:   queue,
+	}, nil
+}
+
+// Run runs audit service 2.0.
+func (service *Service2) Run(ctx context.Context) (err error) {
+	defer mon.Task()(&ctx)(&err)
+	service.log.Info("audit 2.0 is starting up")
+
+	var group errgroup.Group
```

Comment:
You can use here a cycle and a limiter to handle these. It would simplify the code.

Reply:
I added a limiter, but what is the need for the cycle? `process()` already iterates and does polling when the queue is empty.

Reply:
Cycle allows controlling when the worker goes to check the queue. Otherwise, tests will end up sleeping to figure out whether workers have finished or not -- the reason we use cycle in the first place.

Also, you don't want to keep all the workers in a single process to be alive and pinging the queue.

Reply:
i.e. https://github.com/storj/storj/pull/2924/files